Source/BluetoothManager.swift

```diff
@@ -128,70 +120,44 @@ public class BluetoothManager {
+                    return Disposables.create()
                 }
-
-                let scanOperationBox = WeakBox<ScanOperation>()
-
-                // Create new scan which will be processed in a queue
-                let operation = Observable.create { [weak self] (element: AnyObserver<ScannedPeripheral>) -> Disposable in
-                    guard let strongSelf = self else { return Disposables.create() }
-                    // Observable which will emit next element, when peripheral is discovered.
-                    let disposable = strongSelf.delegateWrapper.didDiscoverPeripheral
+                strongSelf.scanDisposable = strongSelf.delegateWrapper.didDiscoverPeripheral
                         .flatMap { [weak self] (cbPeripheral, advertisment, rssi) -> Observable<ScannedPeripheral> in
-                            guard let strongSelf = self else { throw BluetoothError.destroyed }
+                            guard let strongSelf = self else {
+                                throw BluetoothError.destroyed
+                            }
                             let peripheral = Peripheral(manager: strongSelf, peripheral: cbPeripheral)
                             let advertismentData = AdvertisementData(advertisementData: advertisment)
                             return .just(ScannedPeripheral(peripheral: peripheral,
-                                                           advertisementData: advertismentData, rssi: rssi))
-                        }
-                        .subscribe(element)
-
-                    // Start scanning for devices
-                    strongSelf.centralManager.scanForPeripherals(withServices: serviceUUIDs, options: options)
-
-                    return Disposables.create { [weak self] in
-                        guard let strongSelf = self else { return }
-                        // When disposed, stop all scans, and remove scanning operation from queue
-                        strongSelf.centralManager.stopScan()
-                        disposable.dispose()
-                        do { strongSelf.lock.lock(); defer { strongSelf.lock.unlock() }
-                            if let index = strongSelf.scanQueue.index(where: { $0 == scanOperationBox.value! }) {
-                                strongSelf.scanQueue.remove(at: index)
-                            }
+                                    advertisementData: advertismentData, rssi: rssi))
                         }
+                        .subscribe(observer)
+
+                strongSelf.centralManager.scanForPeripherals(withServices: serviceUUIDs, options: options)
+
+                return Disposables.create {
+                    guard let strongSelf = self else { return }
+                    // When disposed, stop scan and dispose scanning
+                    strongSelf.centralManager.stopScan()
+                    do { strongSelf.lock.lock(); defer { strongSelf.lock.unlock() }
```

Comment:
We need to talk more about locking and thread safety of this library.

Reply:
in this issue or later ? :)

Reply:
Yes, we need to talk about it later, before creating task for it. For now it can stay as it is.